opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -883,45 +883,19 @@ namespace Opm {
 
                 // If a production well disallows crossflow and its
                 // (prediction type) rate control is zero, then it is effectively shut.
-                if (!well_ecl.getAllowCrossFlow() && well_ecl.isProducer() && well_ecl.predictionMode()) {
-                    const auto& summaryState = ebosSimulator_.vanguard().summaryState();
-                    const auto prod_controls = well_ecl.productionControls(summaryState);
-
-                    auto is_zero = [](const double x)
-                    {
-                        return std::isfinite(x) && !std::isnormal(x);
-                    };
-
+                if (!well_ecl.getAllowCrossFlow())  {
                     bool zero_rate_control = false;
-                    switch (prod_controls.cmode) {
-                    case Well::ProducerCMode::ORAT:
-                        zero_rate_control = is_zero(prod_controls.oil_rate);
-                        break;
-
-                    case Well::ProducerCMode::WRAT:
-                        zero_rate_control = is_zero(prod_controls.water_rate);
-                        break;
-
-                    case Well::ProducerCMode::GRAT:
-                        zero_rate_control = is_zero(prod_controls.gas_rate);
-                        break;
-
-                    case Well::ProducerCMode::LRAT:
-                        zero_rate_control = is_zero(prod_controls.liquid_rate);
-                        break;
-
-                    case Well::ProducerCMode::RESV:
-                        zero_rate_control = is_zero(prod_controls.resv_rate);
-                        break;
-                    default:
-                        // Might still have zero rate controls, but is pressure controlled.
-                        zero_rate_control = false;
-                        break;
+                    if (well_ecl.isProducer()) {
+                        const auto prod_controls = well_ecl.productionControls(summaryState_);
+                        zero_rate_control = prod_controls.zeroRateLimit();
+                    } else { // injectors
+                        const auto inj_controls = well_ecl.injectionControls(summaryState_);
+                        zero_rate_control = inj_controls.zeroRateLimit();
                     }
 
                     if (zero_rate_control) {
                         // Treat as shut, do not add to container.
-                        local_deferredLogger.info("  Well shut due to zero rate control and disallowing crossflow: " + well_ecl.name());
+                        local_deferredLogger.info(fmt::format("  Well {} gets shut due to zero rate constraint and disallowing crossflow ", well_ecl.name()) );
```

Comment:
We probably does not want to see this message for each time steps for each well.